src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/optvector/SubtypeOnlyFloatVectorToArrayCodec.java

```diff
@@ -0,0 +1,180 @@
+ * between this and {@link
+ * com.datastax.oss.driver.internal.core.type.codec.extras.vector.FloatVectorToArrayCodec} is that
+ * we don't concern ourselves with the dimensionality specified in the input CQL type. This codec
+ * just reads all the bytes, tries to deserislize them consecutively into subtypes and then returns
+ * the result. Serialiation is similar: we take the input array, serialize each element and return
+ * the result.
+ */
+public class SubtypeOnlyFloatVectorToArrayCodec implements TypeCodec<float[]> {
+
+  private static final int ELEMENT_SIZE = 4;
+
+  @NonNull protected final VectorType cqlType;
+  @NonNull protected final GenericType<float[]> javaType;
+
+  private final FloatCodec floatCodec = new FloatCodec();
+
+  private static final SubtypeOnlyFloatVectorToArrayCodec INSTANCE =
+      new SubtypeOnlyFloatVectorToArrayCodec(DataTypes.FLOAT);
+
+  private SubtypeOnlyFloatVectorToArrayCodec(@NonNull DataType subType) {
+    cqlType = new SubtypeOnlyVectorType(Objects.requireNonNull(subType, "subType cannot be null"));
+    javaType = GenericType.of(float[].class);
+  }
+
+  public static TypeCodec<float[]> instance() {
+    return INSTANCE;
+  }
+
+  @NonNull
+  @Override
+  public GenericType<float[]> getJavaType() {
+    return javaType;
+  }
+
+  @NonNull
+  @Override
+  public DataType getCqlType() {
+    return cqlType;
+  }
+
+  @Override
+  public boolean accepts(@NonNull Class<?> javaClass) {
+    return float[].class.equals(javaClass);
+  }
+
+  @Override
+  public boolean accepts(@NonNull Object value) {
+    return value instanceof float[];
+  }
+
+  @Override
+  public boolean accepts(@NonNull DataType value) {
+    if (!(value instanceof VectorType)) {
+      return false;
+    }
+    VectorType valueVectorType = (VectorType) value;
+    return this.cqlType.getElementType().equals(valueVectorType.getElementType());
+  }
+
+  @Nullable
+  @Override
+  public ByteBuffer encode(@Nullable float[] array, @NonNull ProtocolVersion protocolVersion) {
+    if (array == null) {
+      return null;
+    }
+    int length = array.length;
+    int totalSize = length * ELEMENT_SIZE;
+    ByteBuffer output = ByteBuffer.allocate(totalSize);
+    for (int i = 0; i < length; i++) {
+      serializeElement(output, array, i, protocolVersion);
+    }
+    output.flip();
+    return output;
+  }
+
+  @Nullable
+  @Override
+  public float[] decode(@Nullable ByteBuffer bytes, @NonNull ProtocolVersion protocolVersion) {
+    if (bytes == null || bytes.remaining() == 0) {
+      throw new IllegalArgumentException(
+          "Input ByteBuffer must not be null and must have non-zero remaining bytes");
+    }
+    // TODO: Do we want to treat this as an error?  We could also just ignore any extraneous bytes
+    // if they appear.
+    if (bytes.remaining() % ELEMENT_SIZE != 0) {
+      throw new IllegalArgumentException(
+          String.format("Input ByteBuffer should have a multiple of %d bytes", ELEMENT_SIZE));
```

Comment:
Not sure how common this exception is. If this is a common error, should we use Error V2 instead of `IllegalArgumentException`? Have the same question for other `IllegalArgumentException`s. Or all these `IllegalArgumentException`s will be caught eventually and converted to our own errors? I see there is caught in `VectorCodec` and convert it to `ToCQLCodecException`, but not sure if it covers all the cases.

Reply:
This codec is used by Java CQL Driver, it's not a Data API codec. But I think code we use for java driver will have to catch exceptions anyway. Problem itself should not be possible to occur by any external calls; it is more an assertion (and as such not sure there is a way to easily test).